### ISSUE TITLE
one modle train and compile many time, in  pytorch  xla? why  and solution

### ISSUE THREAD

**Issue by ckfgihub** [pytorch/xla#5554]
When the same model is compiled using pytorch Why does the same model produce multiple graphs during training? It is actually very time-consuming to compile multiple graphs.

My question is, when training a model in pytorch xla, why are two graphs generated? Why does the picture after the second step have %31 = "mhlo.tuple"(%16, %22, %25, %26, %arg5, %arg8, %6, %30, %18, %12) % What about the two inputs arg5 and %arg8 as output? What principles or design considerations do designers like pytorch xla base on? If there is any code (framework code in pytorch or xla) that involves this principle or implementation details, can you point it out to me? In addition, is there any optimization method? The purpose is to generate just one picture, reduce the number of compilations, and improve performance. See detailed background below.

The following is my training model:
import torch
import torch.nn as nn
import torch.optim as optim
import torch_xla
import torch_xla.core.xla_model as xm
import os
from torch_xla.utils.utils import get_free_tcp_ports
from torch_xla.amp import syncfree

os.environ['XRT_DEVICE_MAP'] = "CPU:0;/job:localservice/replica:0/task:0/device:XLA_CPU:0"
os.environ['XRT_WORKERS'] = "localservice:0;grpc://localhost:%s" % get_free_tcp_ports()[0]

os.environ["XLA_FLAGS"] = "--xla_dump_to=./simple_model_out"

class SimpleModel(nn.Module):
    def __init__(self, input_size, output_size):
        super(SimpleModel, self).__init__()
        self.linear = nn.Linear(input_size, output_size)

    def forward(self, x):
        return self.linear(x)

input_size = 10
output_size = 5

device = xm.xla_device()
model = SimpleModel(input_size, output_size).to(device)
criterion = nn.MSELoss()
optimizer = optim.SGD(model.parameters(), lr=0.1)

input_data = torch.randn(1, input_size).to(device)
target = torch.randn(1, output_size).to(device)
optimizer = syncfree.SGD(model.parameters(), lr=0.01, momentum=0)

model.train()
found_inf = torch.tensor(0.0).to(device)
for _ in range(10):
    output = model(input_data)
    loss = criterion(output, target)

    optimizer.zero_grad()
    loss.backward()
    optimizer.step(found_inf=found_inf)
    xm.mark_step()

The following is the mhlo converted during the first step of training:
module @SyncTensorsGraph.106 {
  func.func @main(%arg0: tensor<f32>, %arg1: tensor<1x5xf32>, %arg2: tensor<5xf32>, %arg3: tensor<5x10xf32>, %arg4: tensor<1x10xf32>, %arg5: tensor<f32>) -> tuple<tensor<5x10xf32>, tensor<5xf32>, tensor<1x5xf32>, tensor<f32>, tensor<5xf32>, tensor<5x10xf32>, tensor<f32>, tensor<f32>, tensor<5x10xf32>, tensor<5xf32>> {
    %0 = mhlo.constant dense<0.000000e+00> : tensor<f32>
    %1 = mhlo.compare  NE, %arg5, %0 : (tensor<f32>, tensor<f32>) -> tensor<i1>
    %2 = "mhlo.broadcast_in_dim"(%1) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i1>) -> tensor<5x10xi1>
    %3 = "mhlo.transpose"(%arg4) {permutation = dense<[1, 0]> : tensor<2xi64>, xla_shape = "f32[10,1]{0,1}"} : (tensor<1x10xf32>) -> tensor<10x1xf32>
    %4 = mhlo.reshape %arg2 : (tensor<5xf32>) -> tensor<1x5xf32>
    %5 = "mhlo.transpose"(%arg3) {permutation = dense<[1, 0]> : tensor<2xi64>, xla_shape = "f32[10,5]{0,1}"} : (tensor<5x10xf32>) -> tensor<10x5xf32>
    %6 = "mhlo.fusion"(%4, %arg4, %5) ({
    ^bb0(%arg6: tensor<1x5xf32>, %arg7: tensor<1x10xf32>, %arg8: tensor<10x5xf32>):
      %33 = "mhlo.dot"(%arg7, %arg8) {precision_config = [#mhlo<precision DEFAULT>, #mhlo<precision DEFAULT>]} : (tensor<1x10xf32>, tensor<10x5xf32>) -> tensor<1x5xf32>
      %34 = mhlo.add %33, %arg6 : tensor<1x5xf32>
      mhlo.return %34 : tensor<1x5xf32>
    }) {fusion_kind = #mhlo<fusion_kind kLoop>} : (tensor<1x5xf32>, tensor<1x10xf32>, tensor<10x5xf32>) -> tensor<1x5xf32>
    %7 = mhlo.subtract %6, %arg1 : tensor<1x5xf32>
    %8 = mhlo.constant dense<4.000000e-01> : tensor<f32>
    %9 = "mhlo.broadcast_in_dim"(%8) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<f32>) -> tensor<1x5xf32>
    %10 = mhlo.multiply %7, %9 : tensor<1x5xf32>
    %11 = "mhlo.dot"(%3, %10) {precision_config = [#mhlo<precision DEFAULT>, #mhlo<precision DEFAULT>]} : (tensor<10x1xf32>, tensor<1x5xf32>) -> tensor<10x5xf32>
    %12 = "mhlo.transpose"(%11) {permutation = dense<[1, 0]> : tensor<2xi64>, xla_shape = "f32[5,10]{0,1}"} : (tensor<10x5xf32>) -> tensor<5x10xf32>
    %13 = "mhlo.broadcast_in_dim"(%arg0) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<f32>) -> tensor<5x10xf32>
    %14 = mhlo.multiply %12, %13 {xla_shape = "f32[5,10]{0,1}"} : tensor<5x10xf32>
    %15 = mhlo.subtract %arg3, %14 : tensor<5x10xf32>
    %16 = "mhlo.select"(%2, %arg3, %15) : (tensor<5x10xi1>, tensor<5x10xf32>, tensor<5x10xf32>) -> tensor<5x10xf32>
    %17 = "mhlo.broadcast_in_dim"(%1) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i1>) -> tensor<5xi1>
    %18 = mhlo.reduce(%10 init: %0) across dimensions = [0] : (tensor<1x5xf32>, tensor<f32>) -> tensor<5xf32>
     reducer(%arg6: tensor<f32>, %arg7: tensor<f32>)  {
      %33 = mhlo.add %arg6, %arg7 : tensor<f32>
      mhlo.return %33 : tensor<f32>
    }
    %19 = "mhlo.broadcast_in_dim"(%arg0) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<f32>) -> tensor<5xf32>
    %20 = mhlo.multiply %18, %19 : tensor<5xf32>
    %21 = mhlo.subtract %arg2, %20 : tensor<5xf32>
    %22 = "mhlo.select"(%17, %arg2, %21) : (tensor<5xi1>, tensor<5xf32>, tensor<5xf32>) -> tensor<5xf32>
    %23 = mhlo.multiply %7, %7 : tensor<1x5xf32>
    %24 = mhlo.reduce(%23 init: %0) across dimensions = [0, 1] : (tensor<1x5xf32>, tensor<f32>) -> tensor<f32>
     reducer(%arg6: tensor<f32>, %arg7: tensor<f32>)  {
      %33 = mhlo.add %arg6, %arg7 : tensor<f32>
      mhlo.return %33 : tensor<f32>
    }
    %25 = mhlo.constant dense<2.000000e-01> : tensor<f32>
    %26 = mhlo.multiply %24, %25 : tensor<f32>
    %27 = mhlo.not %1 : tensor<i1>
    %28 = mhlo.convert(%27) : (tensor<i1>) -> tensor<f32>
    %29 = mhlo.copy %28 : (tensor<f32>) -> tensor<f32>
    %30 = mhlo.copy %12 {xla_shape = "f32[5,10]{0,1}"} : (tensor<5x10xf32>) -> tensor<5x10xf32>
    %31 = mhlo.copy %18 : (tensor<5xf32>) -> tensor<5xf32>
    %32 = "mhlo.tuple"(%16, %22, %6, %26, %18, %12, %28, %29, %30, %31) {xla_shape = "(f32[5,10]{1,0}, f32[5]{0}, f32[1,5]{1,0}, f32[], f32[5]{0}, /*index=5*/f32[5,10]{0,1}, f32[], f32[], f32[5,10]{0,1}, f32[5]{0})"} : (tensor<5x10xf32>, tensor<5xf32>, tensor<1x5xf32>, tensor<f32>, tensor<5xf32>, tensor<5x10xf32>, tensor<f32>, tensor<f32>, tensor<5x10xf32>, tensor<5xf32>) -> tuple<tensor<5x10xf32>, tensor<5xf32>, tensor<1x5xf32>, tensor<f32>, tensor<5xf32>, tensor<5x10xf32>, tensor<f32>, tensor<f32>, tensor<5x10xf32>, tensor<5xf32>>
    return %32 : tuple<tensor<5x10xf32>, tensor<5xf32>, tensor<1x5xf32>, tensor<f32>, tensor<5xf32>, tensor<5x10xf32>, tensor<f32>, tensor<f32>, tensor<5x10xf32>, tensor<5xf32>>
  }
}

The following is the mhlo graph generated during training after the second step:
module @SyncTensorsGraph.108 {
  func.func @main(%arg0: tensor<f32>, %arg1: tensor<1x5xf32>, %arg2: tensor<5xf32>, %arg3: tensor<5x10xf32>, %arg4: tensor<1x10xf32>, %arg5: tensor<5x10xf32>, %arg6: tensor<f32>, %arg7: tensor<f32>, %arg8: tensor<5xf32>, %arg9: tensor<f32>) -> tuple<tensor<5x10xf32>, tensor<5xf32>, tensor<f32>, tensor<f32>, tensor<5x10xf32>, tensor<5xf32>, tensor<1x5xf32>, tensor<f32>, tensor<5xf32>, tensor<5x10xf32>> {
    %0 = mhlo.constant dense<0.000000e+00> : tensor<f32>
    %1 = mhlo.compare  NE, %arg7, %0 : (tensor<f32>, tensor<f32>) -> tensor<i1>
    %2 = "mhlo.broadcast_in_dim"(%1) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i1>) -> tensor<5x10xi1>
    %3 = "mhlo.transpose"(%arg4) {permutation = dense<[1, 0]> : tensor<2xi64>, xla_shape = "f32[10,1]{0,1}"} : (tensor<1x10xf32>) -> tensor<10x1xf32>
    %4 = mhlo.reshape %arg2 : (tensor<5xf32>) -> tensor<1x5xf32>
    %5 = "mhlo.transpose"(%arg3) {permutation = dense<[1, 0]> : tensor<2xi64>, xla_shape = "f32[10,5]{0,1}"} : (tensor<5x10xf32>) -> tensor<10x5xf32>
    %6 = "mhlo.fusion"(%4, %arg4, %5) ({
    ^bb0(%arg10: tensor<1x5xf32>, %arg11: tensor<1x10xf32>, %arg12: tensor<10x5xf32>):
      %32 = "mhlo.dot"(%arg11, %arg12) {precision_config = [#mhlo<precision DEFAULT>, #mhlo<precision DEFAULT>]} : (tensor<1x10xf32>, tensor<10x5xf32>) -> tensor<1x5xf32>
      %33 = mhlo.add %32, %arg10 : tensor<1x5xf32>
      mhlo.return %33 : tensor<1x5xf32>
    }) {fusion_kind = #mhlo<fusion_kind kLoop>} : (tensor<1x5xf32>, tensor<1x10xf32>, tensor<10x5xf32>) -> tensor<1x5xf32>
    %7 = mhlo.subtract %6, %arg1 : tensor<1x5xf32>
    %8 = mhlo.constant dense<4.000000e-01> : tensor<f32>
    %9 = "mhlo.broadcast_in_dim"(%8) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<f32>) -> tensor<1x5xf32>
    %10 = mhlo.multiply %7, %9 : tensor<1x5xf32>
    %11 = "mhlo.dot"(%3, %10) {precision_config = [#mhlo<precision DEFAULT>, #mhlo<precision DEFAULT>]} : (tensor<10x1xf32>, tensor<1x5xf32>) -> tensor<10x5xf32>
    %12 = "mhlo.transpose"(%11) {permutation = dense<[1, 0]> : tensor<2xi64>, xla_shape = "f32[5,10]{0,1}"} : (tensor<10x5xf32>) -> tensor<5x10xf32>
    %13 = "mhlo.broadcast_in_dim"(%arg0) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<f32>) -> tensor<5x10xf32>
    %14 = mhlo.multiply %12, %13 {xla_shape = "f32[5,10]{0,1}"} : tensor<5x10xf32>
    %15 = mhlo.subtract %arg3, %14 : tensor<5x10xf32>
    %16 = "mhlo.select"(%2, %arg3, %15) : (tensor<5x10xi1>, tensor<5x10xf32>, tensor<5x10xf32>) -> tensor<5x10xf32>
    %17 = "mhlo.broadcast_in_dim"(%1) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i1>) -> tensor<5xi1>
    %18 = mhlo.reduce(%10 init: %0) across dimensions = [0] : (tensor<1x5xf32>, tensor<f32>) -> tensor<5xf32>
     reducer(%arg10: tensor<f32>, %arg11: tensor<f32>)  {
      %32 = mhlo.add %arg10, %arg11 : tensor<f32>
      mhlo.return %32 : tensor<f32>
    }
    %19 = "mhlo.broadcast_in_dim"(%arg0) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<f32>) -> tensor<5xf32>
    %20 = mhlo.multiply %18, %19 : tensor<5xf32>
    %21 = mhlo.subtract %arg2, %20 : tensor<5xf32>
    %22 = "mhlo.select"(%17, %arg2, %21) : (tensor<5xi1>, tensor<5xf32>, tensor<5xf32>) -> tensor<5xf32>
    %23 = mhlo.not %1 : tensor<i1>
    %24 = mhlo.convert(%23) : (tensor<i1>) -> tensor<f32>
    %25 = mhlo.add %arg6, %24 : tensor<f32>
    %26 = mhlo.add %arg9, %24 : tensor<f32>
    %27 = mhlo.multiply %7, %7 : tensor<1x5xf32>
    %28 = mhlo.reduce(%27 init: %0) across dimensions = [0, 1] : (tensor<1x5xf32>, tensor<f32>) -> tensor<f32>
     reducer(%arg10: tensor<f32>, %arg11: tensor<f32>)  {
      %32 = mhlo.add %arg10, %arg11 : tensor<f32>
      mhlo.return %32 : tensor<f32>
    }
    %29 = mhlo.constant dense<2.000000e-01> : tensor<f32>
    %30 = mhlo.multiply %28, %29 : tensor<f32>
    %31 = "mhlo.tuple"(%16, %22, %25, %26, %arg5, %arg8, %6, %30, %18, %12) {xla_shape = "(f32[5,10]{1,0}, f32[5]{0}, f32[], f32[], f32[5,10]{1,0}, /*index=5*/f32[5]{0}, f32[1,5]{1,0}, f32[], f32[5]{0}, f32[5,10]{0,1})"} : (tensor<5x10xf32>, tensor<5xf32>, tensor<f32>, tensor<f32>, tensor<5x10xf32>, tensor<5xf32>, tensor<1x5xf32>, tensor<f32>, tensor<5xf32>, tensor<5x10xf32>) -> tuple<tensor<5x10xf32>, tensor<5xf32>, tensor<f32>, tensor<f32>, tensor<5x10xf32>, tensor<5xf32>, tensor<1x5xf32>, tensor<f32>, tensor<5xf32>, tensor<5x10xf32>>
    return %31 : tuple<tensor<5x10xf32>, tensor<5xf32>, tensor<f32>, tensor<f32>, tensor<5x10xf32>, tensor<5xf32>, tensor<1x5xf32>, tensor<f32>, tensor<5xf32>, tensor<5x10xf32>>
  }
}

**Comment by JackCaoG** [comment]
I think it is that at step 0 some of the optimizer state is empty, and it only got filled up starting step 1. Hence the computation executed in step0 and following steps are slight different.

**Comment by ckfgihub** [comment]
> I think it is that at step 0 some of the optimizer state is empty, and it only got filled up starting step 1. Hence the computation executed in step0 and following steps are slight different.

thank you for your msg.  And, Dynamo, Is it possible to obtain a model into a graph and send it to xla for compilation? That is, it can support dynamic shape and control flow?

**Comment by ckfgihub** [comment]
> I think it is that at step 0 some of the optimizer state is empty, and it only got filled up starting step 1. Hence the computation executed in step0 and following steps are slight different.

Hello, JackCaoG！
I found the existing pytorch xla mechanism returns all liveTensors as the calculation graph, which means that many of the copies from xla to host are not the Tensors that users want to synchronize. The AI ​​training framework transfers them from Returned to host. If the model is relatively large, copying back and forth is also a performance loss. How do you consider this? As shown above, the momentum (%arg5, %arg8） of the optimizer is not used and is directly used as the output. This is an example that does not need to be returned.

**Comment by JackCaoG** [comment]
we support dynamo, you can check out https://github.com/pytorch/xla/blob/master/docs/dynamo.md, but we do not support enrapture control flow and dynamic shape in the graph.

**Comment by JackCaoG** [comment]
output of the graph does not necessary means that we transfer the result back to the host, it just means the value will be computed and they will be lived on the device. Unless you explicitly do `.cpu` call or trying to print the value of a tensor, we won't try to move the value from device to host.
